File: algorithms/bfs.py
import random, pprint

def make_matrix(col, row):
    matrix = [[random.randint(0,10) for x in range(col)] for y in range(row)]
    pprint.pprint(matrix)
    return matrix

def get_neighbors(x,y,matrix,block):
    row = len(matrix)
    col = len(matrix[0])
    n = ((x+1,y), (x-1,y), (x,y-1), (x, y+1), (x-1,y-1), (x+1, y+1), (x-1, y+1), (x+1, y-1))
    neighbors = ((x,y) for x,y in n if 0<= x < row and 0 <= y < col and matrix[x][y] < block)    
    return neighbors

def bfs(matrix, start=None, end=None, block=float('inf')):
    queue = [[start]]
    visited = []

    while queue:
        path = queue.pop(0)
        node = path[-1]
        if node == end:
            return path
        
        x,y = node 
        neighbors = get_neighbors(x,y,matrix,block)
        for cx,cy in neighbors:
            if (cx,cy) not in visited:
                visited.append((cx,cy))
                new_path = list(path)
                new_path.append((cx,cy))
                queue.append(new_path)
    return False
                
    
matrix = make_matrix(5,5)
print(bfs(matrix,(0,0),(4,4)))

File: js/bfs.js
class Node{
  constructor(x,y,v){
    this.x = x;
    this.y = y;
    this.v = v;
    this.searched = false;
    this.parent = false;
  }
}

function make_array(row, col){
  let x = new Array(row);
  for (let i = 0; i < x.length; i++) {
    x[i] = new Array(col)
    for (let j = 0; j < x[i].length; j++){
      x[i][j] =
        new Node(i,j,  Math.floor(Math.random()* 10));
    }
  }
  return x
}

maze = make_array(5,5)
// console.log(maze);
console.log('start')

function bfs(maze, start, end){
  let queue = [[start]]

  while (queue.length){
    let path = queue.shift()
    let node = path[path.length-1]
    console.log('path', node)


    //if end is found
    if (node.v == end){
      console.log('found')
      return 'Found', path
      break
    }

    if (!node.searched){
      node.searched = true
      let neighbors = [[node.x+1,node.y],[node.x-1,node.y],[node.x,node.y+1],[node.x,node.y-1]]
      for(edge of neighbors){
        cx = edge[0]
        cy = edge[1]
        if ( (0 <= cx && cx < maze.length) &&( 0 <= cy && cy < maze[0].length)){
          if (maze[cx][cy].searched == false){
            u = maze[cx][cy]

            let new_path = path.map(x => x)
            console.log('new path', new_path)
            new_path.push(u)
            queue.push(new_path)

          }
        }
      }

    }
  }
return false
}

console.log( bfs(maze, maze[0][0], 5 ))
